### PR TITLE
Win7 SP1 winload.exe revision 23569

### PR DESCRIPTION
**Pull request by jeremitu**
Tested and working. Partly based on <https://github.com/Elbandi/PatchPae2>. Provided exe and Visual Studio 2017 build.

### CODE REVIEW

**Comment by krombel** [comment]
@jeremitu Your patch is not working (at least in my env).
I see the "Starting Windows" screen with the animation and then the screens turns and stays black.
But interestingly I can boot to "Safe Mode" without issues with this patch so I can validate that it uses all of my RAM.

Have you tested with bitlocker enabled? Or any commands to debug that?

**Comment by jeremitu** [comment]
Never used bitlocker. I suppose you have issues with the graphic card, see [comments under original wj32 patch I enhanced](https://wj32.org/wp/2016/02/01/pae-patch-updated-for-windows-10/). 
What graphic card are you using? "Intel HD and Nvidia graphics are known issues." 
These driver problems were the main reasons Microsoft has named when forbidding 4GB+ for 32 bit Windows.

**Comment by krombel** [comment]
Ohh. I have "Intel HD"...
Thanks for you fast answer though

**Comment by jeremitu** [comment]
You might try older video drivers. Good luck!

**Comment by Satoshi64** [comment]
KB4056897 updated winload to 6.1.7601.23992 breaking it again.

**Comment by jeremitu** [comment]
Actually the same patch worked. I opened up the version check.
[The source code is on Github](https://github.com/jeremitu/PatchPae2), the [full story and an exe to test on my website](https://www.jurasz.de/en/jarek/patchpae2).

**Comment by Satoshi64** [comment]
It worked greatly, thank you!
Just had to rollback my video drivers to 332.21 as nvidia refuses to support pae on the newest. Didn't check at exactly which version they stopped.

**Comment by Satoshi64** [comment]
Microsoft has released a new security update, September 11, 2018—KB4457145, which fixes a severe flaw on the way windows handles images. This update has changed:
Ntkrnlpa.exe,6.1.7601.24231,"4,054,192",10-Aug-18,15:45
Ntoskrnl.exe,6.1.7601.24231,"3,961,440",10-Aug-18,15:44
Winload.exe,6.1.7601.24149,"535,616",30-May-18,13:04

**Comment by Satoshi64** [comment]
Microsoft has released a new security update, September 11, 2018—KB4457145, which fixes a severe flaw on the way windows handles images. This update has changed:
Ntkrnlpa.exe,6.1.7601.24231,"4,054,192",10-Aug-18,15:45
Ntoskrnl.exe,6.1.7601.24231,"3,961,440",10-Aug-18,15:44
Winload.exe,6.1.7601.24149,"535,616",30-May-18,13:04
Edit: Patch seems to still work.

**Comment by cnrat** [comment]
> Microsoft has released a new security update, September 11, 2018—KB4457145, which fixes a severe flaw on the way windows handles images. This update has changed:
> Ntkrnlpa.exe,6.1.7601.24231,"4,054,192",10-Aug-18,15:45
> Ntoskrnl.exe,6.1.7601.24231,"3,961,440",10-Aug-18,15:44
> Winload.exe,6.1.7601.24149,"535,616",30-May-18,13:04
> Edit: Patch seems to still work.

Start failed with: MissingOsLoader. It seems MS add verification to winload.exe.

File: PatchPae2/main.c [hunk 1]
```diff
@@ -635,6 +657,127 @@ VOID PatchLoader7601(
...
+    __in PLOADED_IMAGE LoadedImage,
+    __out PBOOLEAN Success
+)
+{
+  // ImgpValidateImageHash - patch to always return 0
+/*
+  .text:004295F7 
+  .text:004295FF
+  .text:004295FF                         loc_4295FF : ; CODE XREF : ImgpValidateImageHash(x, x, x, x, x) + 1Dj
+  .text:004295FF; ImgpValidateImageHash(x, x, x, x, x) + A6j ...
```
Comment (elbandi) [inline PatchPae2/main.c:670]: added an "invalid" char (utf8?) to source code

**Comment by Satoshi64** [comment]
Hello again, i hope this might be useful to anybody still using this in the future.

The last Windows 7 normal & ESU updates since have changed both ntkrnlpa.exe and winload.exe over the years. Right now the updated patch by [jeremitu](https://www.jurasz.de/en/jarek/patchpae2) is able to patch all winloads until version **24538**, any winload.exe above that will instantly fail. Stick to that one.
The tool is also still able to patch all ntkrnlpa.exe kernels until the latest version, 26220. But there is a trap: You will get bsods with error 0x00000100-something while booting on anything above **26174**, so stick with that kernel.

**Comment by anonymoususer16** [comment]
> Hello again, i hope this might be useful to anybody still using this in the future.
> 
> The last Windows 7 normal & ESU updates since have changed both ntkrnlpa.exe and winload.exe over the years. Right now the updated patch by [jeremitu](https://www.jurasz.de/en/jarek/patchpae2) is able to patch all winloads until version **24538**, any winload.exe above that will instantly fail. Stick to that one. The tool is also still able to patch all ntkrnlpa.exe kernels until the latest version, 26220. But there is a trap: You will get bsods with error 0x00000100-something while booting on anything above **26174**, so stick with that kernel.

Version 26909 (2023/12):
edit winloadp.exe (or winload.exe)
Patch region 1
00 04 00 00 22 d8 08 00 10 00  ->  00 04 00 00 1a eb 08 00 10 00

Patch region 2
7c 24 14 8b 44 24 14 5f 5e 5b  ->  7c 24 14 31 c0 90 90 5f 5e 5b

**Pull request merged**